[3/3][Distributed] Remove dist requirement explicit async throws limitation

ktoso opened the pull request:
Depends on:
- #59711
- #59718

### Summary

With this, distributed funcs in protocols are fully functional.

This now correctly implements witnesses and also respects isolation as to when the distributed thunk must be called and when not etc.

This is correct and future proof wrt. any way we'd implement `local` in the future as well, because:

- unlike a concrete distributed actor; even in presence of the "whenLocal" and any potential `local` work, the distributedness _cannot_ be peeled off from a protocol requirement!

- This is because even if we declared the witness as a plain func... the witness is **always allowed to** implement a distributed func requirement using: no effects, async, or throws (or a mix of them)
  - As such... given `any SomeDAProtocol` even if that is a `any _local SomeDAProtocol` can **cannot** ever strip off the implicit async and throws effects from a distributed call on a protocol -- because we don't know (at call sites) how the method was implemented -- and if it was a local actor... we must assume the worst case anyway and `try await` such calls.
  - This does not get into the way of usefulness of the `_local` work -- which's primary purpose is access to NOT-distributed members, which all work as previously.


**Risk:** Low, only impacts distributed funcs in protocols; and is well tested now with runtime (actually calling) tests
**Review:** @DougGregor @xedin 
**Testing:** CI testing, will verify on cluster project manually as well

Resolves rdar://95949498

ktoso commented:
@swift-ci please smoke test

ktoso commented:
@swift-ci please build toolchain

kavon commented on lib/AST/ASTVerifier.cpp:
```diff
@@ -3086,9 +3086,10 @@ class Verifier : public ASTWalker {
 
       // If a decl has the Throws bit set, the ThrowsLoc should be valid,
-      // and vice versa, unless the decl was imported, de-serialized, or
-      // implicit.
+      // and vice versa, unless the decl was imported, de-serialized,
+      // distributed (meaning the throws is implicit), or implicit.
       if (!AFD->isImplicit() &&
+          !AFD->isDistributed() &&
```
This seems fine for now, but makes me think we should rethink the connection between the bit and the location. It gives 4 options in the space of states for the two values, and it seems now at least 3 of them are meaningful. Perhaps in the future we should wrap this complexity behind a method returning an enum to describe these states.

ktoso replied:
<Discussed more and did a rewrite of this>

kavon reviewed (approved):
~After a quick review, I think this looks good.~

kavon reviewed (commented):
github wont let me take back a review 😆

kavon reviewed (changes requested):
Nevermind I'll defer to Doug on this since he is more familiar with the issue.

ktoso commented:
Ok it seems we need to change the approach a little bit -- while outside calls are correct, some inside the actor + extension is not handled well enough wit this.

ktoso committed:
[Distributed] Implement witnesses for sync or non-throw dist reqs

[Distributed] generic and inner test; without one edge case

[Distributed] fix distributed_thunk test; unsure about those extra hops, could remove later

[Distributed] Remove type pretending in getSILFunctionType; it is not needed

It seems our constant replacement in the earlier phases is enough, and
we don't need this trick at all.

[Distributed] Use thunk when calling cross-actor on DA protocols

DougGregor commented on lib/SILGen/SILGenApply.cpp:
```diff
@@ -569,6 +569,20 @@ class Callee {
+    if (constant) {
+      auto func = constant->getFuncDecl();
+      if (func && func->isDistributed()) {
+        fprintf(stderr, "\n[%s:%d] (%s) decl is dist: %d\n", __FILE__, __LINE__, __FUNCTION__, func->isDistributed());
```
Debug dumps

DougGregor commented on include/swift/SIL/SILWitnessVisitor.h:
```diff
@@ -192,6 +194,17 @@ template <class T> class SILWitnessVisitor : public ASTVisitor<T> {
+
+  void addDistributedWitnessMethodsIfRequired(AbstractFunctionDecl *AFD,
+                                              SILDeclRef::Kind kind) {
+    if (!AFD->isDistributed())
```
We need to decide what to do with `distributed async throws` requirements. With this change, they'll have two entries in the witness table. However, we could leave them with just one, which is always the distributed thunk, because (1) the cost of the "is local" check is pretty small once you've already gone through a witness table entry, (2) it's ABI-compatible with your other PR to make `distributed async throws` witnesses point at the distributed thunk.

ktoso replied:
Concretely, this is wrong... or not?

```
    SILDeclRef declRef(AFD, kind);

     // Add another entry for the 'distributed' method:
     asDerived().addMethod(declRef.asDistributed());
```

thr witness table we end up with is:

```
sil_witness_table hidden TheWorker: DistributedWorker module distributed_actor_func_calls_remoteCall_through_generic {
  ... 
  method #DistributedWorker.submit_witness_sync: <Self where Self : DistributedWorker> (isolated Self) -> (Self.WorkItem) -> Self.WorkResult : @$s55distributed_actor_func_calls_remoteCall_through_generic9TheWorkerCAA011DistributedJ0A2aDP19submit_witness_sync5param10WorkResultQz0P4ItemQz_tFTW	// protocol witness for DistributedWorker.submit_witness_sync(param:) in conformance TheWorker

  method #DistributedWorker.submit_witness_sync: <Self where Self : DistributedWorker> (isolated Self) -> (Self.WorkItem) -> Self.WorkResult : @$s55distributed_actor_func_calls_remoteCall_through_generic9TheWorkerCAA011DistributedJ0A2aDP19submit_witness_sync5param10WorkResultQz0P4ItemQz_tFTWTE	// protocol witness for DistributedWorker.submit_witness_sync(param:) in conformance TheWorker
```

so it is the same signature of the function for both actually... is this wrong or right?

Do we instead need to synthesize an AFD based on func here, that has no body, and has `throws` and `async` and add that instead?
```

DougGregor commented on lib/SILGen/SILGenApply.cpp:
```diff
@@ -569,6 +569,20 @@ class Callee {
+
+        if (auto proto = dyn_cast<ProtocolDecl>(func->getDeclContext())) {
+          isDistThunkCall = func->isDistributed() &&
+                            isa<ProtocolDecl>(func->getDeclContext());
```
This whole `&&` is redundant, because we've already checked both `func->isDistributed()` (outer `if`) and that it's in a protocol (inner `if`). But if you take my ABI suggestion above, we'll check for "not `async throws`" here.

ktoso replied:
Ok right the isa is unnecessary, the other part... I'm not sure... I think that'd eventually work if we go all the way with some more refactoring 🤔 Not quite there yet.

DougGregor commented on lib/SILGen/SILGenPoly.cpp:
```diff
@@ -4439,6 +4439,25 @@ void SILGenFunction::emitProtocolWitness(
+            getActorIsolation(witness.getDecl()).isDistributedActor()) {
+
+    if (requirement.isDistributedThunk()) {
+      fprintf(stderr, "[%s:%d] (%s) witness is DIST THUNK!\n", __FILE__, __LINE__, __FUNCTION__);
```
Debug dumps

DougGregor commented on lib/SILGen/SILGenType.cpp:
```diff
@@ -781,6 +793,10 @@ SILFunction *SILGenModule::emitProtocolWitness(
   }
 
+  if (witnessRef.isDistributedThunk()) {
+    nameBuffer = nameBuffer + "TE"; // TE signifies a distributed thunk.
```
This should be handled in `mangleWitnessThunk`.

ktoso replied:
Okey!

ktoso replied:
So this one turns out is not trivial -- AutoDiff also has the issue here: the ASTMangler takes only AST types and we'd need to give it a SILDeclRef... The AutoDiff mangling does it ad hoc here for the same reason I guess. 

It feels like a layering problem to pass SIL stuff to an AST type...? But perhaps it's okey and I should add an optional SILDeclRef there or something similar?

ktoso commented on test/Distributed/Runtime/distributed_actor_func_calls_remoteCall_through_generic_and_inner.swift:
```diff
@@ -0,0 +1,98 @@
+  //  We should understand that greeting is called on self, and not use the thunk witness.
+//  func greetLocal2(name: String) {
+//    print("\(greeting()), \(name)!") // okay, we're on the actor
+//  }
```
This is one last remaining case I could not figure out how to make work yet;

We have to detect that "the call is on self, therefore, do not substitute the constant for the call with the thunk for the witness". But nowhere where we do that, do we have the expr to base it on... In my other `whenLocal` work I did a bunch of improvements around detecting "when a thing is local" like detecting the base of a call... but we still need expressions.

The question about accepting this PR is:
- is this a big enough improvement, to accept this one missing case (it does cause a crash, and the workaround is to make the `greeting()` requirement async throws), that we'd fix later
- or, is this risky and incomplete in a way that we prefer to not merge it.

This also impacts the mangling conflict issue... although I did not manage to switch us over to always passing the `func + asDistributed()` declref rather than the current `thunk + asDistributed()`, but this PR would enable it... rdar://96179707

Open to any plans about how we should tackle this tbh; we can take it slow with this PR as well, not to risk breaking 5.7.

ktoso commented:
Notable last moment removal: https://github.com/apple/swift/pull/59722/commits/fe59943db053d89b37fad48af8d4ab7658b69bef seems we don't need this trickery at all, and the constant swapping is totally enough.

ktoso commented:
@swift-ci please smoke test

ktoso commented:
@swift-ci please smoke test
